File: studyq-backend/models/quiz.py
from datetime import datetime, timezone
from models import db


class Quiz(db.Model):
    __tablename__ = "quizzes"

    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey("users.id"), nullable=False)
    document_id = db.Column(db.Integer, db.ForeignKey("documents.id"), nullable=False)
    created_at = db.Column(
        db.DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    questions = db.relationship(
        "Question", backref="quiz", lazy=True, cascade="all, delete-orphan"
    )
    attempts = db.relationship(
        "Attempt", backref="quiz", lazy=True, cascade="all, delete-orphan"
    )

    def to_dict(self, include_questions=False, hide_answers=True):
        data = {
            "id": self.id,
            "user_id": self.user_id,
            "document_id": self.document_id,
            "created_at": self.created_at.isoformat(),
            "question_count": len(self.questions),
            "attempt_count": len(self.attempts),
        }
        if include_questions:
            data["questions"] = [
                q.to_dict(hide_answer=hide_answers) for q in self.questions
            ]
        return data

File: studyq-backend/models/Quiz.js
const mongoose = require('mongoose');

const quizSchema = new mongoose.Schema({
  user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  document_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Document', required: true },
  created_at: { type: Date, default: Date.now }
});

quizSchema.virtual('questions', {
  ref: 'Question',
  localField: '_id',
  foreignField: 'quiz_id'
});

quizSchema.virtual('attempts', {
  ref: 'Attempt',
  localField: '_id',
  foreignField: 'quiz_id'
});

quizSchema.set('toObject', { virtuals: true });
quizSchema.set('toJSON', { virtuals: true });

quizSchema.methods.toDict = async function(include_questions = false, hide_answers = true) {
  if (!this.populated('questions')) await this.populate('questions');
  if (!this.populated('attempts')) await this.populate('attempts');
  
  const data = {
    id: this._id,
    user_id: this.user_id,
    document_id: this.document_id,
    created_at: this.created_at,
    question_count: this.questions ? this.questions.length : 0,
    attempt_count: this.attempts ? this.attempts.length : 0
  };

  if (include_questions && this.questions) {
    data.questions = this.questions.map(q => q.toDict(hide_answers));
  }

  return data;
};

module.exports = mongoose.model('Quiz', quizSchema);
